azure-iot-device/azure/iot/device/provisioning/registration_client_factory.py

```diff
@@ -0,0 +1,40 @@
+# -------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for
+# license information.
+# --------------------------------------------------------------------------
+"""
+This module provides factory methods for creating clients which would
+communicate with Device Provisioning Service.
+"""
+from .security.sk_security_client import SymmetricKeySecurityClient
+
+from .sk_registration_client import SymmetricKeyRegistrationClient
+from .transport.state_based_mqtt_provider import StateBasedMQTTProvider
+
+
+def create_from_security_client(provisioning_host, security_client, transport_choice):
+    """
+    Creates different types of registration clients which can enable devices to communicate with Device Provisioning
+    Service based on parameters passed.
+    :param provisioning_host: Host running the Device Provisioning Service. Can be found in the Azure portal in the
+    Overview tab as the string Global device endpoint
+    :param security_client: Instance of Security client object which can be either of Symmetric Key, TPM or X.509
+    :param transport_choice: A string representing the transport the user wants
+    :return: A specific registration client based on parameters and validations.
+    """
+    if not provisioning_host:
+        raise ValueError("Provisioning Host must be provided.")
+    transport_choice = transport_choice.lower()
+    if transport_choice == "mqtt":
+        mqtt_state_based_provider = StateBasedMQTTProvider(provisioning_host, security_client)
+        if isinstance(security_client, SymmetricKeySecurityClient):
+            return SymmetricKeyRegistrationClient(mqtt_state_based_provider)
+            # TODO : other instances of security provider can also be checked before creating mqtt
+        else:
+            raise ValueError(
+                "A symmetric key security provider must be provided for instantiating MQTT"
```

Comment:
>A symmetric key security provider must be provided for instantiating MQTT [](start = 17, length = 73)

i'm not sure that's very helpful to the user, who doesn't now what "instantiating MQTT" means.

Reply:
the parent class of MQTT transport which was symmetric key transport has been removed. I think this was for checking that the user is choosing transports those are compatible with the security mechanisms.  the ideal response would be `a symmetric key security provider must be provided for a symmetric key transport`.  or a `a symmetric key security provider must be provided for MQTT`. Since the user has chosen "mqtt" he should know what "mqtt" is...which error message is better ?

---
In reply to: [277849031](https://github.com/Azure/azure-iot-sdk-python-preview/pull/81#discussion_r277849031) [](ancestors = 277849031)